# A Python Quick Start Tutorial
# Lesson \#1: Basic data types and functions
## by Peter Mackenzie-Helnwein
University of Washington, Seattle, WA

pmackenz@uw.edu          
https://www.ce.washington.edu/facultyfinder/peter-mackenzie-helnwein

## Resources

   1. Python Docs: https://docs.python.org/3/
   
   1. Python Tutorial (comprehensive): https://docs.python.org/3/tutorial/index.html
   
   1. Python Library Reference (the nitty-gritty details): https://docs.python.org/3/library/index.html
   
   1. Everything else: http://google.com
   

## Discussing questions from the self-study assignment

This section is to discuss your questions


### Exercise 1

**Theory**:
Stress transformation

$$
\begin{aligned}
\sigma_{x}' &= \sigma_{x} \cos^2\theta + \sigma_{y} \sin^2\theta + 2\tau_{xy} \sin\theta \cos\theta \\
\sigma_{y}' &= \sigma_{x} \sin^2\theta + \sigma_{y} \cos^2\theta - 2\tau_{xy} \sin\theta \cos\theta \\
\tau_{xy}'  &= (\sigma_{y} - \sigma_{x}) \sin\theta \cos\theta + \tau_{xy} (\cos^2\theta - \sin^2\theta) \\
\end{aligned}
$$

**Given**:
Stress state:

$$
\sigma_{x} = 12~ksi~,
~~~
\sigma_{y} = -5.5~ksi~,
~~~
\tau_{xy} = 3.5~ksi 
$$

**Find**:
Components of stress in a rotated coordinate system for $\theta=25^\circ$.

In [2]:
# theta needs to be in radians
# sin and cos are not part of phython

from math import sin, cos, pi, radians

In [3]:
radians(90)

1.5707963267948966

In [4]:
theta  = radians(25)
sigx = 12
sigy = -5.5
tauxy = 3.5

sigxp = sigx*cos(theta)**2 + sigy*sin(theta)**2 + 2*tauxy*sin(theta)*cos(theta)
sigyp = sigx*sin(theta)**2 + sigy*cos(theta)**2 - 2*tauxy*sin(theta)*cos(theta)
taoxyp = (sigx-sigy)*sin(theta)*cos(theta) + tauxy*(cos(theta)**2 - sin(theta)**2)

In [5]:
print("sigma_x, sigma_y = {:12.4f} {:12.4f} ksi".format(sigxp, sigyp))

sigma_x, sigma_y =      11.5555      -5.0555 ksi


### Exercise 2

Convert the calculation from exercise 1 to a function

In [6]:
def stressTransform(sigx, sigy, tauxy, theta):
    sigxp = sigx*cos(theta)**2 + sigy*sin(theta)**2 + 2*tauxy*sin(theta)*cos(theta)
    sigyp = sigx*sin(theta)**2 + sigy*cos(theta)**2 - 2*tauxy*sin(theta)*cos(theta)
    taoxyp = (sigx-sigy)*sin(theta)*cos(theta) + tauxy*(cos(theta)**2 - sin(theta)**2)
    
    return (sigxp,sigyp,taoxyp)

Test your function for 

* $\theta=0$ (what should the result look like?)
* $\theta=25^\circ$
* $\theta=180^\circ$

In [7]:
stress = stressTransform(sigx, sigy, tauxy, theta=25)

In [8]:
print("sig_x = {:8.4f} ksi\nsig_y = {:8.4f} ksi\ntau = {:8.4f} ksi ".format(*stress))#format(stress[0],stress[1],stress[2])) #format(*stress)

sig_x =  10.7751 ksi
sig_y =  -4.2751 ksi
tau =   1.0816 ksi 


### Exercise 3

Let's use this function to study stress transformation.  In order to do so,

1. Create a list of 81 evenly spaced $\theta$-values from $0^\circ$ to $180^\circ$.
2. Compute the transformed stress for each of these values and collect the results in three lists: 

       sigmax_values = ...
       sigmay_values = ...
       tauxy_values = ...

3. print a nice table with 4 columns: $\theta$, $\sigma_x'$, $\sigma_y'$, $\tau_{xy}'$

In [9]:
import numpy as np
#import pandas as pd

theta =[]
sigmax_values = []
sigmay_values = []
tauxy_values = []

for i in range(81):
    th = i*180/80;
    
    theta.append(th)
    (sx,sy,tau) = stressTransform(sigx, sigy, tauxy, th)
    sigmax_values.append(sx)
    sigmay_values.append(sy)
    tauxy_values.append(tau)

#theteL_s.apply(stressTransform(sigx, sigy, tauxy, theta=thetaL)

for item in zip(theta, sigmax_values, sigmay_values, tauxy_values):
    print("{:12.6f}, {:12.6f}, {:12.6f}, {:12.6f},".format(*item))

    0.000000,    12.000000,    -5.500000,     3.500000,
    2.250000,    -2.015819,     8.515819,    -9.291174,
    4.500000,    -3.279975,     9.779975,     0.417081,
    6.750000,    11.268801,    -4.768801,     9.115336,
    9.000000,     6.399316,     0.100684,    -4.260030,
   11.250000,    -6.096526,    12.596526,    -7.319343,
   13.500000,     4.041101,     2.458899,     7.345804,
   15.750000,    12.263005,    -5.763005,     4.222414,
   18.000000,    -1.340908,     7.840908,    -9.125938,
   20.250000,    -3.827516,    10.327516,    -0.374995,
   22.500000,    10.824730,    -4.324730,     9.284033,
   24.750000,     7.134074,    -0.634074,    -3.539075,
   27.000000,    -5.962223,    12.462223,    -7.791989,
   29.250000,     3.249722,     3.250278,     6.824112,
   31.500000,    12.462340,    -5.962340,     4.915000,
   33.750000,    -0.633567,     7.133567,    -8.896235,
   36.000000,    -4.325061,    10.825061,    -1.164423,
   38.250000,    10.327149,    -3.827149,     9.

In [ ]:
# version two: use pandas to apply function to a dataframe


### Exercise 4: Explore plotting in python

For this, check out https://matplotlib.org/3.1.1/gallery/lines_bars_and_markers/simple_plot.html#sphx-glr-gallery-lines-bars-and-markers-simple-plot-py

I copied the demo code below so you can modify it to plot $(\sigma_{x}'/\tau_{xy}')$ and  $(\sigma_{y}'/-\tau_{xy}')$ curves from the data collected in `sigmax_values`, `sigmay_values`, `tauxy_values`

In [ ]:
import matplotlib
import matplotlib.pyplot as plt
import numpy as np

# Data for plotting
t = np.arange(0.0, 2.0, 0.01)
s = 1 + np.sin(2 * np.pi * t)

fig, ax = plt.subplots()
ax.plot(t, s)

ax.set(xlabel='time (s)', ylabel='voltage (mV)',
       title='About as simple as it gets, folks')
ax.grid()

# I comment writing the figure to file ... makes this quicker
# fig.savefig("test.png")
plt.show()

**Remark**: This was just to wet your appetite! We will talk much more about `numpy` and `matplotlib` next week.

## Exercise 5: List versus dictionary

You could express the stress state as

1. three independent variables - the way we did it thus far

2. a list of three values: `stress = `$[\sigma_x, \sigma_y, \tau_{xy}]$ and replace

        sigma_x --> stress[0]
        sigma_y --> stress[2]
        tau_xy  --> stress[3]
        
3. a dictionary 

        stress = {'sigx': 12.0, 'sigy': -5.5, 'tau':3.50}
        
   and replace

        sigma_x --> stress['sigx']
        sigma_y --> stress['sigy']
        tau_xy  --> stress['tau']

**Discuss**:

1. Pros and cons for coding the equations
2. Pros and cons for coding the function(s)
3. How woud code and function change when switching from 2D to 3D (6 components instead of 3 components)

Feel free to implement an alternative form of stressTrannsform() that looks like this:

In [ ]:
def stressTransform(stress, theta):
    # YOUR CODE
    # stress2 is a list or dictionary, just as stress is
    
    return stress2

## Homework questions

We can all learn from your questions!